File: v4/rust/iprange-livedb/src/recovery/page_set.rs
#[cfg(any(unix, windows))]
use std::mem;
use std::mem::size_of;
#[cfg(any(unix, windows))]
use std::path::PathBuf;

use crate::contract::MetaV4;
use crate::error::{Error, Result};

#[cfg(any(unix, windows))]
use super::scratch::{residue_error, Scratch, ScratchFile, ScratchSlot, HEADER_SIZE};
use super::{RecoveryBudget, ScratchCleanup};

const EMPTY: u64 = 0;
const MIN_SLOTS: usize = 8;
const MAX_LOAD_NUMERATOR: usize = 3;
const MAX_LOAD_DENOMINATOR: usize = 4;

enum Slots {
    Heap(Vec<u64>),
    #[cfg(any(unix, windows))]
    File(FileSlots),
}

#[cfg(any(unix, windows))]
struct FileSlots {
    file: ScratchFile,
    slots: usize,
}

#[cfg(any(unix, windows))]
#[derive(Clone)]
struct Fallback {
    directory: PathBuf,
    source: MetaV4,
    max_bytes: u64,
    max_files: u32,
    max_open_files: u32,
    wanted_slots: usize,
}

/// Sparse page set which migrates to authorized fixed-slot scratch when needed.
pub(crate) struct PageSet {
    slots: Slots,
    len: usize,
    #[cfg(any(unix, windows))]
    fallback: Option<Fallback>,
    #[cfg(any(unix, windows))]
    scratch: Option<Scratch>,
}

pub(crate) struct PageSetFailure {
    pub(crate) cause: Error,
    pub(crate) cleanup: Option<ScratchCleanup>,
}

impl PageSet {
    pub(crate) fn for_recovery(
        max_heap_bytes: u64,
        expected_pages: u64,
        source: MetaV4,
        budget: &RecoveryBudget,
    ) -> Result<Self> {
        #[cfg(any(unix, windows))]
        let fallback = budget.scratch_directory.as_ref().map(|directory| Fallback {
            directory: directory.clone(),
            source,
            max_bytes: budget.max_scratch_bytes,
            max_files: budget.max_scratch_files,
            max_open_files: budget.max_open_files,
            wanted_slots: wanted_slots(expected_pages),
        });
        #[cfg(not(any(unix, windows)))]
        let fallback = {
            let _ = (source, budget);
            None
        };
        Self::allocate(max_heap_bytes, expected_pages, fallback)
    }

    #[cfg(any(unix, windows))]
    fn allocate(
        max_heap_bytes: u64,
        expected_pages: u64,
        fallback: Option<Fallback>,
    ) -> Result<Self> {
        let affordable =
            usize::try_from(max_heap_bytes / size_of::<u64>() as u64).unwrap_or(usize::MAX);
        if affordable < MIN_SLOTS && fallback.is_none() {
            return Err(Error::BudgetExceeded("recovery page-ownership table"));
        }
        let wanted = wanted_slots(expected_pages);
        let slots = if wanted <= affordable {
            wanted
        } else {
            floor_power_of_two(affordable)
        };
        let values = heap_slots(slots)?;
        Ok(Self {
            slots: Slots::Heap(values),
            len: 0,
            fallback,
            scratch: None,
        })
    }

    #[cfg(not(any(unix, windows)))]
    fn allocate(max_heap_bytes: u64, expected_pages: u64, _fallback: Option<()>) -> Result<Self> {
        let affordable =
            usize::try_from(max_heap_bytes / size_of::<u64>() as u64).unwrap_or(usize::MAX);
        if affordable < MIN_SLOTS {
            return Err(Error::BudgetExceeded("recovery page-ownership table"));
        }
        let slots = wanted_slots(expected_pages).min(floor_power_of_two(affordable));
        Ok(Self {
            slots: Slots::Heap(heap_slots(slots)?),
            len: 0,
        })
    }

    /// Returns `true` for a new page and `false` for a prior claim.
    pub(crate) fn insert(&mut self, page: u32) -> Result<bool> {
        if exceeds_load(self.len + 1, self.slot_count()) {
            #[cfg(any(unix, windows))]
            self.migrate()?;
            if exceeds_load(self.len + 1, self.slot_count()) {
                return Err(Error::BudgetExceeded("recovery page-ownership table"));
            }
        }
        let encoded = u64::from(page) + 1;
        let mask = self.slot_count() - 1;
        let mut index = hash(page) & mask;
        loop {
            match self.read(index)? {
                EMPTY => {
                    self.write(index, encoded)?;
                    self.len += 1;
                    return Ok(true);
                }
                value if value == encoded => return Ok(false),
                _ => index = (index + 1) & mask,
            }
        }
    }

    pub(crate) fn retained_bytes(&self) -> u64 {
        match &self.slots {
            Slots::Heap(slots) => slots.len() as u64 * size_of::<u64>() as u64,
            #[cfg(any(unix, windows))]
            Slots::File(_) => 0,
        }
    }

    pub(crate) fn reset(&mut self) -> Result<()> {
        self.len = 0;
        match &mut self.slots {
            Slots::Heap(slots) => slots.fill(EMPTY),
            #[cfg(any(unix, windows))]
            Slots::File(_) => self.reset_file()?,
        }
        Ok(())
    }

    #[allow(clippy::result_large_err)]
    pub(crate) fn finish(
        self,
        result: Result<()>,
    ) -> std::result::Result<Option<ScratchCleanup>, PageSetFailure> {
        #[cfg(any(unix, windows))]
        {
            finish_cleanup(self.cleanup(), result)
        }
        #[cfg(not(any(unix, windows)))]
        {
            drop(self);
            result.map(|()| None).map_err(|cause| PageSetFailure {
                cause,
                cleanup: None,
            })
        }
    }

    #[cfg(any(unix, windows))]
    pub(crate) fn take_scratch(&mut self) -> Option<Scratch> {
        self.fallback = None;
        self.scratch.take()
    }

    #[cfg(any(unix, windows))]
    pub(crate) fn release(mut self, scratch: &mut Scratch) -> Option<ScratchSlot> {
        match mem::replace(&mut self.slots, Slots::Heap(Vec::new())) {
            Slots::Heap(_) => None,
            Slots::File(slots) => {
                let slot = scratch.attach(slots.file);
                Some(slot)
            }
        }
    }

    #[cfg(any(unix, windows))]
    pub(crate) fn create_scratch_file(&mut self, length: u64) -> Result<ScratchFile> {
        let scratch = self.ensure_scratch()?;
        let slot = scratch.create()?;
        scratch.resize(slot, length)?;
        scratch.detach(slot)
    }

    #[cfg(any(unix, windows))]
    pub(crate) fn cleanup(mut self) -> Option<ScratchCleanup> {
        if let Some(mut scratch) = self.scratch.take() {
            let _ = self.release(&mut scratch);
            Some(scratch.cleanup())
        } else {
            None
        }
    }

    fn slot_count(&self) -> usize {
        match &self.slots {
            Slots::Heap(slots) => slots.len(),
            #[cfg(any(unix, windows))]
            Slots::File(slots) => slots.slots,
        }
    }

    fn read(&self, index: usize) -> Result<u64> {
        match &self.slots {
            Slots::Heap(slots) => Ok(slots[index]),
            #[cfg(any(unix, windows))]
            Slots::File(slots) => slots.read(index),
        }
    }

    fn write(&mut self, index: usize, value: u64) -> Result<()> {
        match &mut self.slots {
            Slots::Heap(slots) => {
                slots[index] = value;
                Ok(())
            }
            #[cfg(any(unix, windows))]
            Slots::File(slots) => slots.write(index, value),
        }
    }

    #[cfg(any(unix, windows))]
    fn migrate(&mut self) -> Result<()> {
        if matches!(self.slots, Slots::File(_)) {
            return Ok(());
        }
        let fallback = self
            .fallback
            .clone()
            .ok_or(Error::BudgetExceeded("recovery page-ownership table"))?;
        let slots = self.file_slots(&fallback)?;
        if exceeds_load(self.len + 1, slots) {
            return Err(Error::BudgetExceeded("recovery page-ownership table"));
        }
        let output = self.create_file_slots(slots)?;
        self.slots = Slots::File(output);
        Ok(())
    }

    #[cfg(any(unix, windows))]
    fn create_file_slots(&mut self, slots: usize) -> Result<FileSlots> {
        let length = table_length(slots)?;
        let scratch = self.ensure_scratch()?;
        let slot = scratch.create()?;
        scratch.resize(slot, length)?;
        let file = scratch.detach(slot)?;
        let mut output = FileSlots { file, slots };
        self.copy_claims(&mut output)?;
        Ok(output)
    }

    #[cfg(any(unix, windows))]
    fn ensure_scratch(&mut self) -> Result<&mut Scratch> {
        if self.scratch.is_none() {
            let fallback = self
                .fallback
                .as_ref()
                .ok_or(Error::BudgetExceeded("recovery scratch"))?;
            self.scratch = Some(start_scratch(fallback)?);
        }
        Ok(self.scratch.as_mut().expect("scratch was initialized"))
    }

    #[cfg(any(unix, windows))]
    fn file_slots(&self, fallback: &Fallback) -> Result<usize> {
        let available = match &self.scratch {
            Some(scratch) => scratch.remaining_bytes(),
            None => {
                let reserve = if fallback.max_files >= 2 {
                    HEADER_SIZE
                } else {
                    0
                };
                fallback
                    .max_bytes
                    .checked_sub(reserve)
                    .ok_or(Error::BudgetExceeded("recovery page-ownership scratch"))?
            }
        };
        fallback.file_slots(available)
    }

    #[cfg(any(unix, windows))]
    fn copy_claims(&self, output: &mut FileSlots) -> Result<()> {
        let Slots::Heap(values) = &self.slots else {
            return Ok(());
        };
        for &value in values.iter().filter(|&&value| value != EMPTY) {
            output.insert(value)?;
        }
        Ok(())
    }

    #[cfg(any(unix, windows))]
    fn reset_file(&mut self) -> Result<()> {
        let Slots::File(slots) = &self.slots else {
            unreachable!("file reset requires file slots")
        };
        let scratch = self
            .scratch
            .as_mut()
            .expect("file page set retains its scratch attempt");
        let slot = slots.file.slot();
        scratch.reset(slot)?;
        scratch.resize(slot, table_length(slots.slots)?)?;
        Ok(())
    }
}

#[cfg(any(unix, windows))]
#[allow(clippy::result_large_err)]
fn finish_cleanup(
    cleanup: Option<ScratchCleanup>,
    result: Result<()>,
) -> std::result::Result<Option<ScratchCleanup>, PageSetFailure> {
    let Some(cleanup) = cleanup else {
        return result.map(|()| None).map_err(|cause| PageSetFailure {
            cause,
            cleanup: None,
        });
    };
    if cleanup.clean() {
        return match result {
            Ok(()) => Ok(Some(cleanup)),
            Err(cause) => Err(PageSetFailure {
                cause,
                cleanup: Some(cleanup),
            }),
        };
    }
    let cause = Error::CleanupIncomplete {
        cause: Box::new(match result {
            Ok(()) => Error::Corrupt("recovery scratch cleanup is incomplete"),
            Err(cause) => cause,
        }),
        cleanup: Box::new(residue_error(&cleanup)),
    };
    Err(PageSetFailure {
        cause,
        cleanup: Some(cleanup),
    })
}

#[cfg(any(unix, windows))]
impl FileSlots {
    fn read(&self, index: usize) -> Result<u64> {
        let mut bytes = [0; 8];
        self.file.read(slot_offset(index)?, &mut bytes)?;
        Ok(u64::from_le_bytes(bytes))
    }

    fn write(&self, index: usize, value: u64) -> Result<()> {
        self.file.write(slot_offset(index)?, &value.to_le_bytes())
    }

    fn insert(&mut self, encoded: u64) -> Result<()> {
        let page = u32::try_from(encoded - 1)
            .map_err(|_| Error::Corrupt("recovery page claim is invalid"))?;
        let mask = self.slots - 1;
        let mut index = hash(page) & mask;
        loop {
            if self.read(index)? == EMPTY {
                return self.write(index, encoded);
            }
            index = (index + 1) & mask;
        }
    }
}

#[cfg(any(unix, windows))]
impl Fallback {
    fn file_slots(&self, available: u64) -> Result<usize> {
        let table_bytes = available
            .checked_sub(HEADER_SIZE)
            .ok_or(Error::BudgetExceeded("recovery page-ownership scratch"))?;
        let affordable =
            usize::try_from(table_bytes / size_of::<u64>() as u64).unwrap_or(usize::MAX);
        let slots = self.wanted_slots.min(floor_power_of_two(affordable));
        if slots < MIN_SLOTS {
            return Err(Error::BudgetExceeded("recovery page-ownership scratch"));
        }
        Ok(slots)
    }
}

#[cfg(any(unix, windows))]
fn start_scratch(fallback: &Fallback) -> Result<Scratch> {
    Scratch::start(
        &fallback.directory,
        fallback.source,
        fallback.max_bytes,
        fallback.max_files,
        fallback.max_open_files,
    )
}

fn heap_slots(slots: usize) -> Result<Vec<u64>> {
    if slots < MIN_SLOTS {
        return Ok(Vec::new());
    }
    let mut values = Vec::new();
    values
        .try_reserve_exact(slots)
        .map_err(|_| Error::BudgetExceeded("recovery page-ownership table"))?;
    values.resize(slots, EMPTY);
    Ok(values)
}

fn wanted_slots(expected_pages: u64) -> usize {
    let wanted = expected_pages
        .saturating_mul(MAX_LOAD_DENOMINATOR as u64)
        .div_ceil(MAX_LOAD_NUMERATOR as u64)
        .max(MIN_SLOTS as u64);
    usize::try_from(wanted)
        .ok()
        .and_then(usize::checked_next_power_of_two)
        .unwrap_or(1usize << (usize::BITS - 1))
}

fn exceeds_load(len: usize, slots: usize) -> bool {
    slots == 0
        || len.saturating_mul(MAX_LOAD_DENOMINATOR) > slots.saturating_mul(MAX_LOAD_NUMERATOR)
}

#[cfg(any(unix, windows))]
fn table_length(slots: usize) -> Result<u64> {
    (slots as u64)
        .checked_mul(size_of::<u64>() as u64)
        .and_then(|bytes| bytes.checked_add(HEADER_SIZE))
        .ok_or(Error::ArithmeticOverflow(
            "recovery page-ownership scratch length",
        ))
}

#[cfg(any(unix, windows))]
fn slot_offset(index: usize) -> Result<u64> {
    (index as u64)
        .checked_mul(size_of::<u64>() as u64)
        .and_then(|bytes| bytes.checked_add(HEADER_SIZE))
        .ok_or(Error::ArithmeticOverflow(
            "recovery page-ownership scratch offset",
        ))
}

fn floor_power_of_two(value: usize) -> usize {
    if value == 0 {
        0
    } else {
        1usize << (usize::BITS - 1 - value.leading_zeros())
    }
}

fn hash(page: u32) -> usize {
    let mut value = u64::from(page);
    value ^= value >> 30;
    value = value.wrapping_mul(0xbf58_476d_1ce4_e5b9);
    value ^= value >> 27;
    value = value.wrapping_mul(0x94d0_49bb_1331_11eb);
    (value ^ (value >> 31)) as usize
}

#[cfg(test)]
#[path = "page_set_tests.rs"]
mod tests;

File: v4/go/internal/recovery/page_set.go
package recovery

// Recovery page-ownership set (Rust recovery/page_set.rs): one sparse
// u64 page set which proves every recovered output page is written
// exactly once. The heap table is a power-of-two open hash table with
// the Rust load cap; when the authorized scratch budget is supplied,
// the load cap migrates the table into fixed-slot file storage inside
// the scratch directory (table at offset 128, one 8-byte claim per
// slot), and the scratch attempt travels with the set into the
// external sort and the cleanup terminal.

import (
	"math"
	"math/bits"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/validation"
)

const (
	pageSetEmpty              = uint64(0)
	pageSetMinSlots           = 8
	pageSetMaxLoadNumerator   = 3
	pageSetMaxLoadDenominator = 4
	slotBytes                 = 8
)

// pageSet is the sparse page-ownership table of one recovery output
// (Rust PageSet with the Heap or File slot storage and the optional
// scratch fallback).
type pageSet struct {
	heap     []uint64
	file     *fileSlots
	len      int
	fallback *pageFallback
	scratch  *scratch
}

// pageFallback is the authorized-scratch migration description of one
// page set (Rust Fallback).
type pageFallback struct {
	directory    string
	source       format.Meta
	maxBytes     uint64
	maxFiles     uint32
	maxOpenFiles uint32
	wantedSlots  int
}

// fileSlots is the fixed-slot file table of one migrated page set
// (Rust FileSlots: the detached scratch file and the slot count).
type fileSlots struct {
	file  scratchFile
	slots int
}

// newPageSet allocates the heap table for a recovery budget (Rust
// PageSet::allocate heap arm): the affordable slot count is the heap
// budget divided by the slot size, the wanted count derives from the
// expected page count at the load cap, and the table is the smaller
// power of two. Without a scratch fallback a budget below the minimum
// slot count refuses; with one, the tiny table migrates on load.
func newPageSet(maxHeapBytes uint64, expectedPages uint64, fallback *pageFallback) (*pageSet, error) {
	affordable := maxHeapBytes / slotBytes
	if affordable > math.MaxInt {
		affordable = math.MaxInt
	}
	if affordable < pageSetMinSlots && fallback == nil {
		return nil, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery page-ownership table"}
	}
	wanted := wantedSlots(expectedPages)
	slots := wanted
	if int(affordable) < wanted {
		slots = floorPowerOfTwo(int(affordable))
	}
	values, err := heapSlots(slots)
	if err != nil {
		return nil, err
	}
	return &pageSet{heap: values, fallback: fallback}, nil
}

// forRecovery builds the page set of one recovery construction (Rust
// PageSet::for_recovery): the budget's scratch shape is accepted and
// validated, the fallback captures the scratch directory facts, and
// the heap table is the smaller power of two.
func forRecovery(maxHeapBytes uint64, expectedPages uint64, source format.Meta, budget *RecoveryBudget) (*pageSet, error) {
	if err := budget.validate(); err != nil {
		return nil, err
	}
	var fallback *pageFallback
	if budget.ScratchDirectory != "" {
		fallback = &pageFallback{
			directory:    budget.ScratchDirectory,
			source:       source,
			maxBytes:     budget.MaxScratchBytes,
			maxFiles:     budget.MaxScratchFiles,
			maxOpenFiles: budget.MaxOpenFiles,
			wantedSlots:  wantedSlots(expectedPages),
		}
	}
	return newPageSet(maxHeapBytes, expectedPages, fallback)
}

// insert claims one page, reporting whether it is new (Rust
// PageSet::insert): the load cap is enforced before the probe, the
// cap triggers the scratch migration when a fallback exists, and a
// still-exceeding table refuses with the BudgetExceeded class.
func (p *pageSet) insert(page uint32) (bool, error) {
	if exceedsLoad(p.len+1, p.slotCount()) {
		if err := p.migrate(); err != nil {
			return false, err
		}
		if exceedsLoad(p.len+1, p.slotCount()) {
			return false, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery page-ownership table"}
		}
	}
	encoded := uint64(page) + 1
	mask := uint64(p.slotCount() - 1)
	index := hashU32(page) & mask
	for {
		value, err := p.readSlot(int(index))
		if err != nil {
			return false, err
		}
		switch {
		case value == pageSetEmpty:
			if err := p.writeSlot(int(index), encoded); err != nil {
				return false, err
			}
			p.len++
			return true, nil
		case value == encoded:
			return false, nil
		default:
			index = (index + 1) & mask
		}
	}
}

// claim walks one graph edge through the ownership set (Rust
// PageSet::claim): depth and bounds refusals carry their deterministic
// reasons, an already-claimed page is the cycle or alias class from
// the walk path, and a successful claim records the page on the path.
func (p *pageSet) claim(page uint32, pageCount uint64, path []uint32, depth int) (bool, validation.ValidationReason, error) {
	if depth >= len(path) {
		return false, validation.ReasonTreeLevelInvalid, nil
	}
	if page < 2 || uint64(page) >= pageCount {
		return false, validation.ReasonPageOutOfBounds, nil
	}
	newClaim, err := p.insert(page)
	if err != nil {
		return false, 0, err
	}
	if !newClaim {
		reason := validation.ReasonPageAlias
		if containsPage(path[:depth], page) {
			reason = validation.ReasonTreeCycle
		}
		return false, reason, nil
	}
	path[depth] = page
	return true, 0, nil
}

// slotCount is the current table capacity (Rust PageSet::slot_count).
func (p *pageSet) slotCount() int {
	if p.file != nil {
		return p.file.slots
	}
	return len(p.heap)
}

// readSlot reads one claim (Rust PageSet::read: heap or fixed-slot
// file).
func (p *pageSet) readSlot(index int) (uint64, error) {
	if p.file != nil {
		return p.file.read(index)
	}
	return p.heap[index], nil
}

// writeSlot stores one claim (Rust PageSet::write).
func (p *pageSet) writeSlot(index int, value uint64) error {
	if p.file != nil {
		return p.file.write(index, value)
	}
	p.heap[index] = value
	return nil
}

// retainedBytes is the heap retained by the table (Rust
// PageSet::retained_bytes: the file table retains no heap).
func (p *pageSet) retainedBytes() uint64 {
	if p.file != nil {
		return 0
	}
	return uint64(len(p.heap)) * slotBytes
}

// reset clears every claim (Rust PageSet::reset).
func (p *pageSet) reset() error {
	p.len = 0
	if p.file != nil {
		return p.resetFile()
	}
	for i := range p.heap {
		p.heap[i] = pageSetEmpty
	}
	return nil
}

// pageSetFailure is the failing terminal of one page set (Rust
// PageSetFailure): the cause and the scratch cleanup of the attempted
// migration.
type pageSetFailure struct {
	cause   error
	cleanup *scratchCleanup
}

// finish folds the operation result through the page-set terminal
// (Rust PageSet::finish over finish_cleanup): the retained scratch
// attempt cleans before the result is delivered, and an unclean
// cleanup folds into the CleanupIncomplete class with the cleanup
// ledger attached.
func (p *pageSet) finish(result error) *pageSetFailure {
	var cleanup *scratchCleanup
	if p.scratch != nil {
		scratch := p.scratch
		p.scratch = nil
		scratch.attachSlotIfFile(p)
		cleanup = scratch.cleanup()
	}
	if cleanup == nil {
		return &pageSetFailure{cause: result, cleanup: nil}
	}
	if cleanup.clean() {
		return &pageSetFailure{cause: result, cleanup: cleanup}
	}
	return &pageSetFailure{cause: cleanupIncompleteError(result, cleanup), cleanup: cleanup}
}

// attachSlotIfFile returns the detached file slot of the page set to
// its scratch owner (Rust release arm during cleanup).
func (s *scratch) attachSlotIfFile(p *pageSet) {
	if p.file != nil {
		s.attach(p.file.file)
	}
}

// takeScratch detaches the scratch attempt for the external sort
// (Rust PageSet::take_scratch: the fallback is consumed and the
// attempt moves to the caller).
func (p *pageSet) takeScratch() *scratch {
	p.fallback = nil
	scratch := p.scratch
	p.scratch = nil
	return scratch
}

// release returns the page set's file storage to its scratch attempt
// (Rust PageSet::release): the detached file slot is re-attached and
// reported, the storage is reset to an empty heap table.
func (p *pageSet) release(scratch *scratch) *scratchSlot {
	var slot *scratchSlot
	if p.file != nil {
		value := scratch.attach(p.file.file)
		slot = &value
		p.file = nil
	}
	p.heap = nil
	return slot
}

// createScratchFile establishes one detached scratch file of the
// requested length (Rust PageSet::create_scratch_file: the table
// arm of Tables::allocate).
func (p *pageSet) createScratchFile(length uint64) (scratchFile, error) {
	scratch, err := p.ensureScratch()
	if err != nil {
		return scratchFile{}, err
	}
	slot, err := scratch.create()
	if err != nil {
		return scratchFile{}, err
	}
	if err := scratch.resize(slot, length); err != nil {
		return scratchFile{}, err
	}
	return scratch.detach(slot), nil
}

// migrate moves the heap claims into fixed-slot file storage (Rust
// PageSet::migrate); without a fallback the BudgetExceeded class is
// the Rust non-posix arm.
func (p *pageSet) migrate() error {
	if p.file != nil {
		return nil
	}
	if p.fallback == nil {
		return &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery page-ownership table"}
	}
	fallback := *p.fallback
	slots, err := p.fileSlotsOf(&fallback)
	if err != nil {
		return err
	}
	if exceedsLoad(p.len+1, slots) {
		return &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery page-ownership table"}
	}
	output, err := p.createFileSlots(slots)
	if err != nil {
		return err
	}
	p.file = output
	return nil
}

// fileSlotsOf computes the affordable fixed-slot count (Rust
// PageSet::file_slots over Fallback::file_slots).
func (p *pageSet) fileSlotsOf(fallback *pageFallback) (int, error) {
	var available uint64
	if p.scratch != nil {
		available = p.scratch.remainingBytes()
	} else {
		reserve := uint64(0)
		if fallback.maxFiles >= 2 {
			reserve = scratchHeaderSize
		}
		available = fallback.maxBytes - reserve
		if available > fallback.maxBytes {
			return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery page-ownership scratch"}
		}
	}
	return fallback.fileSlots(available)
}

// createFileSlots establishes the detached fixed-slot table (Rust
// PageSet::create_file_slots: the table extent, the scratch file, and
// the copy of every existing claim).
func (p *pageSet) createFileSlots(slots int) (*fileSlots, error) {
	length, err := tableLength(slots)
	if err != nil {
		return nil, err
	}
	scratch, err := p.ensureScratch()
	if err != nil {
		return nil, err
	}
	slot, err := scratch.create()
	if err != nil {
		return nil, err
	}
	if err := scratch.resize(slot, length); err != nil {
		return nil, err
	}
	file := scratch.detach(slot)
	output := &fileSlots{file: file, slots: slots}
	for _, value := range p.heap {
		if value == pageSetEmpty {
			continue
		}
		if err := output.insert(value); err != nil {
			return nil, err
		}
	}
	return output, nil
}

// ensureScratch starts the scratch attempt on first use (Rust
// PageSet::ensure_scratch).
func (p *pageSet) ensureScratch() (*scratch, error) {
	if p.scratch == nil {
		if p.fallback == nil {
			return nil, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery scratch"}
		}
		scratch, err := scratchStart(p.fallback.directory, p.fallback.source, p.fallback.maxBytes, p.fallback.maxFiles, p.fallback.maxOpenFiles)
		if err != nil {
			return nil, err
		}
		p.scratch = scratch
	}
	return p.scratch, nil
}

// resetFile clears the fixed-slot table (Rust PageSet::reset_file:
// reset then re-establish the table extent).
func (p *pageSet) resetFile() error {
	if p.file == nil {
		panic("file reset requires file slots")
	}
	scratch, err := p.ensureScratch()
	if err != nil {
		return err
	}
	slot := p.file.file.slot()
	if err := scratch.reset(slot); err != nil {
		return err
	}
	length, err := tableLength(p.file.slots)
	if err != nil {
		return err
	}
	return scratch.resize(slot, length)
}

// pageFallback.fileSlots computes the affordable slot count from the
// available bytes (Rust Fallback::file_slots: the header first, the
// slot count at the floor power of two, the load cap, and the minimum
// slot count).
func (f *pageFallback) fileSlots(available uint64) (int, error) {
	tableBytes := available - scratchHeaderSize
	if tableBytes > available {
		return 0, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery page-ownership scratch"}
	}
	affordable := tableBytes / slotBytes
	if affordable > math.MaxInt {
		affordable = math.MaxInt
	}
	slots := f.wantedSlots
	if int(affordable) < slots {
		slots = floorPowerOfTwo(int(affordable))
	}
	if slots < pageSetMinSlots {
		return 0, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "recovery page-ownership scratch"}
	}
	return slots, nil
}

// tableLength is the complete fixed-slot table extent (Rust
// table_length: the header plus the slot bytes).
func tableLength(slots int) (uint64, error) {
	bytes, ok := checkedMul(uint64(slots), slotBytes)
	if !ok {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery page-ownership scratch length"}
	}
	total, ok := checkedAdd(bytes, scratchHeaderSize)
	if !ok {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery page-ownership scratch length"}
	}
	return total, nil
}

// slotOffset is the byte offset of one fixed-slot claim (Rust
// slot_offset: the header plus the slot index).
func slotOffset(index int) (uint64, error) {
	bytes, ok := checkedMul(uint64(index), slotBytes)
	if !ok {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery page-ownership scratch offset"}
	}
	total, ok := checkedAdd(bytes, scratchHeaderSize)
	if !ok {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery page-ownership scratch offset"}
	}
	return total, nil
}

// fileSlots.read reads one claim (Rust FileSlots::read).
func (s *fileSlots) read(index int) (uint64, error) {
	var bytes [8]byte
	offset, err := slotOffset(index)
	if err != nil {
		return 0, err
	}
	if err := s.file.read(offset, bytes[:]); err != nil {
		return 0, err
	}
	return format.U64(bytes[:]), nil
}

// fileSlots.write stores one claim (Rust FileSlots::write).
func (s *fileSlots) write(index int, value uint64) error {
	var bytes [8]byte
	format.PutU64(bytes[:], value)
	offset, err := slotOffset(index)
	if err != nil {
		return err
	}
	return s.file.write(offset, bytes[:])
}

// fileSlots.insert claims one page probing from its hash (Rust
// FileSlots::insert; the migration copy arm).
func (s *fileSlots) insert(encoded uint64) error {
	page := uint32(encoded - 1)
	if encoded == 0 || uint64(page)+1 != encoded {
		return &format.Error{Code: format.CodeFormatInvalid, Detail: "recovery page claim is invalid"}
	}
	mask := uint64(s.slots - 1)
	index := hashU32(page) & mask
	for {
		value, err := s.read(int(index))
		if err != nil {
			return err
		}
		if value == pageSetEmpty {
			return s.write(int(index), encoded)
		}
		index = (index + 1) & mask
	}
}

// heapSlots allocates the zeroed table (Rust heap_slots; the
// allocation is bounded by the budget through the affordable slot
// count).
func heapSlots(slots int) ([]uint64, error) {
	if slots < pageSetMinSlots {
		return []uint64{}, nil
	}
	return make([]uint64, slots), nil
}

// wantedSlots sizes the table for the expected page count at the load
// cap (Rust wanted_slots: saturating ceil(pages * 4 / 3), at least
// the minimum, rounded up to a power of two; the saturating clamp to
// the widest usize maps to the top bit of the platform int).
func wantedSlots(expectedPages uint64) int {
	wanted := uint64(pageSetMinSlots)
	if expectedPages <= math.MaxUint64/pageSetMaxLoadDenominator {
		product := expectedPages * pageSetMaxLoadDenominator
		wanted = (product + pageSetMaxLoadNumerator - 1) / pageSetMaxLoadNumerator
		if wanted < pageSetMinSlots {
			wanted = pageSetMinSlots
		}
	}
	if wanted > uint64(maxInt) {
		return maxInt
	}
	return nextPowerOfTwo(int(wanted))
}

// exceedsLoad reports whether one more claim breaches the load cap
// (Rust exceeds_load: len * 4 > slots * 3).
func exceedsLoad(length, slots int) bool {
	return slots == 0 || uint64(length)*pageSetMaxLoadDenominator > uint64(slots)*pageSetMaxLoadNumerator
}

// floorPowerOfTwo is the largest power of two at most value (Rust
// floor_power_of_two).
func floorPowerOfTwo(value int) int {
	if value <= 0 {
		return 0
	}
	return 1 << (bits.Len(uint(value)) - 1)
}

// nextPowerOfTwo rounds one value up to a power of two (Rust
// checked_next_power_of_two).
func nextPowerOfTwo(value int) int {
	if value <= 1 {
		return 1
	}
	return 1 << bits.Len(uint(value-1))
}

// containsPage reports whether the walk path contains one page (Rust
// path[..depth].contains).
func containsPage(path []uint32, page uint32) bool {
	for _, entry := range path {
		if entry == page {
			return true
		}
	}
	return false
}

// hashU32 is the page hash of the recovery tables module (Rust
// tables::hash_u32, the splitmix-style finalizer over the raw page
// number).
func hashU32(value uint32) uint64 {
	v := uint64(value)
	v ^= v >> 30
	v *= 0xbf58476d1ce4e5b9
	v ^= v >> 27
	v *= 0x94d049bb133111eb
	return v ^ (v >> 31)
}
